obs-webrtc: Improve WHIP compliance

Sean-Der opened the pull request:
### Description
Improves WHIP compliance around the handling of Location headers.

If the location header isn't present we don't allow the session to start. We also now support Absolute and Relative Location headers. 

This cherry-picks https://github.com/obsproject/obs-studio/pull/9306 and then adds additional checks

### Motivation and Context
I want to make sure we are as compliant as possible. If we ship a release with compliance issues we will have to support it forever. 

### How Has This Been Tested?
Tested against 
* AWS IVS
* Broadcast Box
* Cloudflare

### Types of changes
- Bug fix (non-breaking change which fixes an issue)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Sean-Der commented:
@notr1ch This covers all the compliance issues you raised in https://github.com/obsproject/obs-studio/pull/9306 I believe!

RytoEX commented:
This appears to contain #9306, which should be noted in the PR description.

Sean-Der commented:
@RytoEX updated the description. Does that read ok?

RytoEX commented:
> @RytoEX updated the description. Does that read ok?

Yes, that's fine.

DDRBoxman commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -347,28 +349,55 @@ bool WHIPOutput::Connect()
+	CURLU *h = curl_url();
+	auto last_location_header = location_headers.back();
+
+	if (last_location_header.find("http") != 0) {
```
TIL http finally supports relative redirects in Location headers, neat.

Can we stick a comment here saying that this is what were doing here?

Also should we check that it's a valid url path too or is the protocol check good enough?

Sean-Der replied:
Done! Added a comment on Relative vs Absolute URL handling

DDRBoxman commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -347,28 +349,55 @@ bool WHIPOutput::Connect()
+		return false;
+	}
+
+	CURLU *h = curl_url();
```
Nit-picky but can we give this a more descriptive name? my head keeps reading the h as header further down, heh

Sean-Der replied:
Done! Changed name to `url_builder` I think that better communicates its purpose.

Sean-Der commented:
@DDRBoxman thank you for the review! I addressed both comments, mind looking again?

RytoEX commented:
To those looking to merge: I would like to see #9306 approved and merged first if this is meant to stack on top of it. If this is meant to supersede #9306 entirely, then I would prefer that it is closed first.

Sean-Der commented:
@RytoEX This is meant to supersede. Can you please #9306 ? This PR modifies the commit that was cherry-picked

notr1ch commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -347,28 +349,58 @@ bool WHIPOutput::Connect()
+	curl_easy_getinfo(c, CURLINFO_REDIRECT_COUNT, &redirect_count);
+
+	if (location_headers.size() < static_cast<size_t>(redirect_count) + 1) {
 		do_log(LOG_WARNING,
```
This should probably be `LOG_ERROR` now.

Sean-Der replied:
Changed to `LOG_ERROR`

notr1ch commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -347,28 +349,58 @@ bool WHIPOutput::Connect()
+	CURLUcode rc = curl_url_get(url_builder, CURLUPART_URL, &url,
+				    CURLU_NO_DEFAULT_PORT);
+	if (rc) {
+		do_log(LOG_WARNING,
+		       "WHIP server provided a invalid resource URL via the Location header");
```
Also here.

Sean-Der replied:
Changed to `LOG_ERROR`

Sean-Der commented:
@notr1ch I have updated all the `LOG_WARNING` messages appropriately. In cases where streaming wouldn't start because of an issue I have changed them.

Can I get another review please? thank you

Sean-Der commented:
@RytoEX Yes can you please close it, thank you!
